File: src/Asm.MooBank.Models/Account.cs
namespace Asm.MooBank.Models;

public partial record Account
{
    public required Guid Id { get; set; }

    public required string Name { get; set; }

    public string? Description { get; set; }

    public DateTimeOffset BalanceDate { get; set; } = DateTimeOffset.Now;

    public DateOnly? LastTransaction { get; set; }

    public required decimal CurrentBalance { get; set; }

    public decimal CalculatedBalance { get; set; }

    public string? AccountType { get; set; }

    public Guid? AccountGroupId { get; set; }
}

File: src/Asm.MooBank.Web/MooBankApp/src/models/Account.ts
import { VirtualAccount } from "./VirtualAccount";

export const AccountTypes = ["None", "Transaction", "Savings", "Credit", "Mortgage", "Superannuation"] as AccountType[];
export type AccountType = "None" | "Transaction" | "Savings" | "Credit" | "Mortgage" | "Superannuation";

export const AccountControllers = ["Manual", "Virtual", "Import"] as AccountController[];
export type AccountController = "Manual" | "Virtual" | "Import";

export type accountId = string;

export interface AccountBase {
    id: accountId;
    name: string;
    description?: string;
    currentBalance: number;
}

export interface Account extends AccountBase {
    calculatedBalance: number;
    balanceDate: Date;
    lastTransaction?: string;
    accountType: AccountType;
    accountGroupId: string;
    controller: AccountController;
    importerTypeId?: number;
    virtualAccountRemainingBalance?: number;
    isPrimary?: boolean;
    shareWithFamily: boolean;
    institutionId: number;
    virtualAccounts: VirtualAccount[];
}

export const emptyAccount : Account = {
    id: "",
    name: "",
    currentBalance: 0,
    balanceDate: new Date(),
    accountType: "None",
    controller: "Manual",
    accountGroupId: "",
    calculatedBalance: 0,
    shareWithFamily: false,
    institutionId: 0,
    virtualAccounts: [],
}

export interface ImportAccount {
    importerTypeId: number;
}
